### PR TITLE
chore(performance): `file-source` testing and benchmarks

### PR DESCRIPTION
**Pull request by blt**
This PR commit is the start of a process to address #6730 by way of implementation of [RFC 3480](https://github.com/timberio/vector/blob/master/rfcs/2020-10-15-3480-file-source-rework.md). The major changes introduced here are criterion benchmarks for an internal component of the `file-source` crate and modification of the testing regime to make use of Quickcheck. The results are nothing ground shaking: the new model tests do not turn up any flaw in the `read_until...` function and the benchmarks show it to be reasonably fast. That's good. Now we know. The more recent indications from [vector-test-harness](https://github.com/timberio/vector-test-harness/pull/74#pullrequestreview-611111679) are that our issue centers not on any individually expensive operation but on the cooperative, fair scheduling of this crate. Approaches to that are beyond the scope of this PR. 

Signed-off-by: Brian L. Troutwine <brian@troutwine.us>

### CODE REVIEW

File: lib/file-source/src/file_watcher.rs [hunk 1]
```diff
@@ -247,12 +251,26 @@ fn null_reader() -> impl BufRead {
...
+/// are read then the reader is polled until `delim` is found and the results
+/// are discarded. Else, the result is written into `buf`.
+///
+/// The return is unusual. In the Err case this function has not written into
+/// `buf` and the caller should not examine its contents. In the Ok case if the
+/// inner value is None the caller should retry the call as the buffering read
+/// hit the end of the buffer but did not find a `delim` yet, indicating that
+/// we've sheered a write or that there were no bytes available in the `reader`
+/// and the `reader` was very sure about it. If the inner value is Some the
+/// interior `usize` is the number of bytes written into `buf`.
```
Comment (lukesteensen) [inline lib/file-source/src/file_watcher.rs:267]: There's more context on why it's in this unusual state in the description of #4089. Very much a classic "we'll clean this up soon" 🙂

Comment (blt) [reply lib/file-source/src/file_watcher.rs:267]: A good friend of mine was prone to say "there's nothing permanent like temporary". :joy:

File: lib/file-source/src/buffer.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+                        byte_range: (previous_offset..global_index),
+                    });
+                    previous_offset = global_index + 1;
+                }
+                global_index += 1;
+            }
+        }
+
+        // Our model is only concerned with the first valid delimiter in the
+        // chunk stream. As such, discover which that it and record it
```
Comment (ktff) [inline lib/file-source/src/buffer.rs:155]:
```suggestion
        // chunk stream. As such, discover which that is and record it
```

**Review by ktff — COMMENTED** [review]
The errors seem to originate from upgrading `bytes` from `0.5` to `1.0`, other than that this looks good.

File: lib/file-source/src/buffer.rs (hunk 2 above)
Comment (blt) [reply lib/file-source/src/buffer.rs:155]: Hmm, "which that" is an awkward construction in itself. Good call-out.

**Comment by blt** [comment]
@ktff indeed. It looks like upgrading `bytes` is non-trivial and should be sequenced after #6513 since there's no pressing need to get `bytes` upgraded and most of the build issues are related to tokio concerns. I'll flip back to 0.5.

**Review by ktff — APPROVED** [review]

File: lib/file-source/src/file_watcher/tests/common.rs [hunk 3]
```diff
@@ -0,0 +1 @@
+
```
Comment (lukesteensen) [inline lib/file-source/src/file_watcher/tests/common.rs:1]: Do we need this file?

Comment (blt) [reply lib/file-source/src/file_watcher/tests/common.rs:1]: We sure do not. Just junk I accidentally left in-tree.

**Comment by pablosichert** [comment]
A merge conflict has summoned me, just came to confirm: `bytes` `1.x` is included in https://github.com/timberio/vector/pull/5872.

**Pull request merged**